### ISSUE TITLE
Please use cmake hunter instead of ExternalProject for dependent libraries

### ISSUE THREAD

**Issue by ned14** [Refinitiv/Elektron-SDK#116]
https://docs.hunter.sh/en/latest/ is very popular in large multinational corporations for managing common open source dependencies. This is because one fixes in stone a particular combination of versions of dependencies, all of which are known to work well with one another thanks to cmake hunter having done the validation, and that specific combination of versions of all dependencies can be then applied across the whole project equally. Builds are perfectly repeatable, and the local toolchain is always used. Problems such as binary incompatible mixed version editions of CURL or LZ4 corrupting memory, or links failing due to colliding or incompatible symbols, no longer occur.

Right now Elektron-SDK does provide some flexibility for injecting externally managed editions of Elektron-SDK's dependencies, but that support is brittle and hard to use. You currently manually implement what cmake hunter does for you by manually pinning versions of subdependencies, and having cmake fetch their source archives and build them via ExternalProject.

cmake hunter is a much better solution to this problem. Apart from saving you the maintenance hassle, it lets consumers of Elektron-SDK configure custom builds of dependencies e.g. with AddressSanitiser enabled, or with different ABI options. cmake hunter is also pluggable and extensible, indeed if Elektron-SDK implemented itself as a cmake hunter dependency, that would be fabulous and make our consumption of your SDK fit into our workflows and release processes. cmake hunter is portable to anywhere cmake works, so this solves all platforms at once.

Note that the original maintainer of cmake hunter has recently retired. Others have taken over maintenance. It can be now found at https://github.com/cpp-pm/hunter.

**Comment by ned14** [comment]
Thanks for your feedback. Firstly, this is the current cmake wrapper we are using to build Elektron-SDK using cmake hunter:

```
cmake_minimum_required(VERSION 3.8)
if(POLICY CMP0074)
  cmake_policy(SET CMP0074 NEW)  # find_package() ought to search _ROOT variables
endif()
include("../master_project/cmake/HunterVersion.cmake")
project(Elektron)
set(gtest_USE_INSTALLED On)           ## GTest::GTest
set(zlib_USE_INSTALLED On)            ## ZLIB::ZLIB
set(lz4_USE_INSTALLED On)             ## LZ4::LZ4
set(libxml2_USE_INSTALLED On)         ## LibXml2::LibXml2
set(curl_USE_INSTALLED On)            ## CURL::libcurl
hunter_add_package(GTest)
find_package(GTest CONFIG REQUIRED)
hunter_add_package(ZLIB)
find_package(ZLIB CONFIG REQUIRED)
if(NOT TARGET ZLIB::ZLIB)
  set_target_properties(
      ZLIB::zlib
      PROPERTIES
      IMPORTED_GLOBAL True
  )
  add_library(ZLIB::ZLIB ALIAS ZLIB::zlib)
endif()
hunter_add_package(lz4)
find_package(lz4 CONFIG REQUIRED)
hunter_add_package(libxml2)
find_package(libxml2 CONFIG REQUIRED)
set(libxml2_version "2.9.7")
hunter_add_package(CURL)
find_package(CURL CONFIG REQUIRED)
set(BUILD_SHARED_LIBS Off)
set(BUILD_ETA_APPLICATIONS Off)
set(BUILD_UNIT_TESTS Off)
set(BUILD_UNIT_TESTS Off)
file(RELATIVE_PATH build_dir "${CMAKE_BINARY_DIR}/.." "${CMAKE_BINARY_DIR}")
set(RCDEV_REL_Ema_LIB_DIR "../../elektron_wrapper/${build_dir}/lib")
set(RCDEV_REL_Eta_LIB_DIR "../../elektron_wrapper/${build_dir}/lib")
add_subdirectory("../elektron" "elektron")
```
It works well.

> Many of the packages used by Hunter are forked from the official repositories. Since we do not have control over the Hunter repositories, it is an unacceptable risk to pull from a third party repository without proper review over each version.

In past roles I simply run an automated diff between the Hunter repo and the corresponding version of the official repo. If the diff's change from last-known-good, it gets flagged for human review and whitelisting of SHAs.

The SHA specified to `HunterGate()` locks into stone a specific set of git SHAs corresponding to specific commits in the Hunter repos. It would be quite difficult to rewrite git history to introduce new code for a pinned commit whilst retaining the same SHA.

> In addition, the hunter repository for curl, for example, is version 7.60, and ESDK requires 7.64. To fix this issue, we would need to backport the curl 7.64 changes into the hunter repository, and then update the hunter scripting with an internal-only cmake script, which exposes the same support concerns as above.

cmake hunter's packages are a *collaborative* effort. If you need a newer version, **you** do the merge from HEAD into the cmake hunter fork and submit a github PR. Such PRs tend to get merged within a day or two, sometimes only hours.

You would be doing such merge work *anyway*, and it's entirely fair that you contribute back to others what you yourself take.

> Almost all of the publicly available superbuild/dependency management frameworks such as Hunter, tomviz, or BASIS assume that all dependencies are static code bases that will not change during development. As the internal Refinitiv project does have a combination of open source and non-open sourced code(the binaries from those are released in the binary pack tarballs), we have an internal requirement of being able to modify and easily rebuild multiple dependencies at once. The dependency framework exposed in ESDK is a part of a fuller internal framework that helps us achieve this.

I don't think that you looked into cmake hunter sufficiently. It has extensive facilities for injecting custom or temporary editions of specific dependencies, or having them built with custom options and flags. You can also pin or gate versions of specific dependencies e.g. hold back onto OpenSSL 1.0.

Such functionality is mainly driven by the needs of the large corporations which need to integrate multiple third party SDKs each using their own, incompatible, clashing versions of open source dependencies.

As an example, I spent this entire week reworking your SDK to use cmake hunter, because your choice of LZ4 version clashed with the choice of LZ4 version by another SDK, and both your SDKs get statically linked into our final binaries. This diverted me from more productive work. Your SDK is not the only SDK in existence, and your choices here cost your downstream customers inconvenience in order for you to internally have more convenience.

If cmake hunter doesn't appeal to you, can I ask for you to refactor how you discover your open source library dependencies to be more configurable please?

Specifically, I'd like to be able to invoke your cmake with a targets import file which specifies the imported targets for the open source projects your SDK uses. Your cmake would include that targets import file, and use those targets instead of doing anything on its own. If no targets import file is supplied, then sure, download and build your own.

**Comment by MitchellKato** [comment]
>Your SDK is not the only SDK in existence, and your choices here cost your downstream customers inconvenience in order for you to internally have more convenience.

You are correct, and this is the intended use case for <target>_USE_INSTALLED tokens, so clients can override with their versions.
